### ISSUE TITLE
Transcribe() WER results differ from validation results for Fastconformer Hybrid RNNT-CTC

### ISSUE THREAD

**Issue by gabitza-tech** [NVIDIA/NeMo#7449]
**Describe the bug**
Hello and, first of all, thank you for your work! I have the following problem:

While training a Fastconformer Hybrid RNNT-CTC model, I perform multi-validation on several datasets at the end of each epoch. The reported results on these datasets at validation, greatly differ from the results obtained when using **speech_to_text_eval.py** (or inherently the **transcribe_speech.py** script). They differ by a margin of +-2% WER on almost all datasets, some having smaller differences than others. (for default RNNT decoding) When I trained the normal fastconformer_transducer model, I haven't experienced such big gaps in WER between training and offline inference.

I have tried providing the model as it is to the _model_path_ argument or firstly changed the decoder to rnnt and then load it, but I obtained same results in both cases. I am also aware of the _dither_ argument in the _preprocessor_ from other issues, and tried changing it to **0**, but didn't obtain any change in the results. 

I have trained the model using bfloat16, but at inference i can't use bfloat16 as I am using a **CPU only** machine, might this be a reason for such a great difference? Should I use a different script for Hybrid models? 

Differences for dataset 1: (10.93% train vs 11.68% inf)
![ssc-eval1_merge](https://github.com/NVIDIA/NeMo/assets/53811655/21a53f1c-1d6d-4578-b7fc-9fd5dc5fc778)

Differences for dataset 2: (12.84% train vs 14.08% inf)
![ssc-eval2_merge](https://github.com/NVIDIA/NeMo/assets/53811655/a297c56d-b9de-40c8-96dd-abbc6307ca1e)

Differences for dataset 3: (3.149% train vs 2.94% inf)
![rsc_eval_merge](https://github.com/NVIDIA/NeMo/assets/53811655/ac0a84fc-de77-47e4-b92e-5fdc452ac6c9)

**Steps/Code to reproduce bug**

Config file used for training:

https://drive.google.com/file/d/1S4eLgDMSsRoLYYxqey-kPOTOHjZfwCOA/view?usp=drive_link

Script used:
```
python3 speech_to_text_eval.py model_path=<path_to_hybrid_model> dataset_manifest=<path_to_manifest_file> output_filename=<path_to_out_json>
```

The **transcribe_speech.py** and **speech_to_text_eval.py** are unchanged. The **config for the fastconformer hybrid** model is almost the same as the normal one.

Unfortunately, I can't provide the datasets I trained/evaluated on, but I also tried to reproduce the experiments on Common Voice Spanish 7.0 dataset (cv-corpus-7.0-2021-07-21), but with a much smaller difference between the results..

Results on dev-set: (6.96% train vs 6.86% inf)
![cv_dev_merge](https://github.com/NVIDIA/NeMo/assets/53811655/ffedb14f-bb04-4f43-bbec-4df1cc662d37)

**Expected behavior**

Same or similar results ( difference of 0.1-0.2% WER ) between validation and offline inference (when using transcribe_speech.py) for consistent training.

**Environment details**

- pytorch-lightning             1.9.4
- torch                         2.1.0a0+fe05266
- torch-tensorrt                1.4.0.dev0
- torchaudio                    2.1.0
- torchmetrics                  0.11.4
- torchvision                   0.15.0a0
- python         3.8.10

**Environment overview (please complete the following information)**

 - Tested offline inference using the main branch code of NeMo repo at 18.09.2023 .

**Additional context**

Trained on A100 gpu with bf16 precision, inference on CPU only. I am also using online noise and speed augmentations while training. I initialiazed the Hybrid model from a transducer pretrained model, but I obtained the same behavior (big differences in WER) while also initializing from a hybrid model.

**Comment by titu1994** [comment]
What is the ddp size when training ? Due to the way ddp works, samples on both train and val set get duplicated on order to have sufficient samples across all ranks of GPUs. This shows up as the small difference between train and offline one gpu eval - especially in MCV. 

The score on the file is a close value to the final result, but for academic results you have to use a small batch size (8/32) and single GPU only in order to get accurate results on your dataset. 

The duplication issue is more prominent if you use large number of nodes, and number of samples in the dataset is not cleanly divisible. 

What you can try is to copy your train script, and replace trainer.fit() with trainer.validate(). It will run the eval under ddp conditions and so should get numbers similar to what you see during train. 

However, it's odd that only hybrid model should this but not transducer model (at least not as strong). So it might be something else

**Comment by gabitza-tech** [comment]
Hey @titu1994, thank you for your response!

We are using 1 A100 gpu with a batch size of 128 and accumulated_grad_batch of 8. For validation we are using a batch size of 8. At inference using **speech_to_text_eval.pyI**, we only use CPU. For our use case, we can't use a GPU at inference.

Is it a problem that we are using the CPU at inference? Tried using it with AMP 16 bit precision, but got the same result. The MCV spanish case is ok, as it is only a 0.1% difference, but the difference on the other datasets start getting bigger.

We also trained some fastconformer RNNT models that achieved good results and don't present such big differences. (but they were trained in different conditions: 2 quadro p5000 with batch size=16 and accum_grad=16) 


Do you have a recommendation on how to achieve similar results using the CPU and the standard transcribe scripts? Or how we should train more consistently?

**Comment by titu1994** [comment]
CPU is fine for inference. AMP and other things will not work for CPU though, they are cuda flags. 

If it's one GPU, its odd to see such difference in results during training vs inference. 

How long do you train - epochs or steps ? And with what LR? Perhaps the model hasn't finished training ? 

> with different optimization parameter

The hyper params of hybrid model are close to that of rnnt models usually. However it can be that the ctc loss is making optimization difficult and therefore loss stays high throughout training. 

You can reduce the effect of ctc by using less than 0.3 weight for ctc loss. 0.1 should also be fine. 

> Do you have a recommendation on how to achieve similar results using the CPU and the standard transcribe scripts? Or how we should train more consistently? Why does ddp affect results when I am running training on a single node?

To answer these questions I need to know your dataset size, avg and max seq length, model size, the entire training config basically. Scheduler, optimizer hp, spec Aug, and so many more factors all affect training. 

Also ddp does not affect results on single GPU but will affect results on more than one GPU - even If it's single node of 4 or 8 gpus.
